### PR TITLE
Fix response ci header

### PR DESCRIPTION
**Pull request by yunstanford**
fix https://github.com/channelcat/sanic/issues/1242

### CODE REVIEW

File: sanic/server.py [hunk 1]
```diff
@@ -39,25 +39,6 @@ class Signal:
     stopped = False
 
 
-class CIDict(dict):
```
Comment (yunstanford) [inline sanic/server.py]: move this class to avoid circular import.

**Comment by vltr** [comment]
Just a minor (?) question: on some other issue<sup>[1]</sup>, it was mentioned (I think optionally) to use [multidict](https://github.com/aio-libs/multidict) for header handling. The question is: why using a custom `CIDict` implementation instead of `multidict`? I know that one of the answers would be to "not add another dependency to the project", but it can be optional (since `CIDict` is already on the Sanic codebase), right? I created a [gist with a simple benchmark](https://gist.github.com/vltr/7965fac5e7db795c803137377ffd3674) parsing some headers and accessing them using `CIDict` and `multidict.CIMultiDict`. Even though my benchmark code may seem a bit superficial (with a lot of copy and paste from related projects), it can provide some initial insights regarding the speed gain by using `multidict.CIMultiDict` ...

<sup>**[1]**</sup> issues (in here) are becoming a bit chaotic to keep track, I think there must be at least two other issues and another one or two PRs regarding this question about headers :confused:

**Comment by yunstanford** [comment]
This pr is just quick fix, i'm fine with converting to `multidict` or sth. proper.

**Comment by vltr** [comment]
Oh, thanks @yunstanford ! I was just confused about this PR (the "whys"). I thought this was the "de facto" solution for #1242 and others :wink:

**Comment by vltr** [comment]
@r0fls:

> @vltr please motion to close any duplicate issues, so that the creator or one of the maintainers can do so.

I have no ties with this project (AFAIK, but I wish I could), so, why will I be blocking anything? :confused:

> @vltr can you share the results of the benchmark?

Yes, it is a (very) simple benchmark. Specs:

```
$ uname -a        
Linux ---- 4.16.16-1-ck-piledriver #1 SMP PREEMPT Sat Jun 16 09:07:54 EDT 2018 x86_64 GNU/Linux

$ hwinfo --short       
cpu:                                                            
                       AMD FX-8320E Eight-Core Processor, 3722 MHz
                       AMD FX-8320E Eight-Core Processor, 3712 MHz
                       AMD FX-8320E Eight-Core Processor, 3715 MHz
                       AMD FX-8320E Eight-Core Processor, 3772 MHz
                       AMD FX-8320E Eight-Core Processor, 3500 MHz
                       AMD FX-8320E Eight-Core Processor, 2994 MHz
                       AMD FX-8320E Eight-Core Processor, 3270 MHz
                       AMD FX-8320E Eight-Core Processor, 3316 MHz
keyboard:
  /dev/input/event18   Microsoft Wired Keyboard 600 (model 1576)
mouse:
  /dev/input/mice      Microsoft Comfort Mouse 3000
monitor:
                       LG ELECTRONICS LG ULTRAWIDE
                       LG ELECTRONICS 23MP65
                       LG ELECTRONICS 23MP65
graphics card:
                       Gigabyte Radeon R9 380X
sound:
                       ATI Tonga HDMI Audio [Radeon R9 285/380]
                       ATI SBx00 Azalia (Intel HDA)
                       Logitech G430 Surround Sound Gaming Headset
storage:
                       ATI SB7x0/SB8x0/SB9x0 SATA Controller [AHCI mode]
                       OCZ RD400/400A SSD
                       ATI SB7x0/SB8x0/SB9x0 IDE Controller
network:
  lan0                 Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller
network interface:
  lan0                 Ethernet network interface
  lo                   Loopback network interface
disk:
  /dev/sda             SanDisk SDSSDA48
  /dev/nvme0n1         OCZ RD400/400A SSD
partition:
  /dev/sda1            Partition
  /dev/sda5            Partition
  /dev/sda6            Partition
  /dev/sda7            Partition
  /dev/sda8            Partition
  /dev/nvme0n1p1       Partition
  /dev/nvme0n1p2       Partition
  /dev/nvme0n1p5       Partition
usb controller:
                       ATI SB7x0/SB8x0/SB9x0 USB EHCI Controller
                       ATI SB7x0/SB8x0/SB9x0 USB OHCI0 Controller
                       Etron EJ188/EJ198 USB 3.0 Host Controller
                       ASMedia USB Controller
                       ATI SB7x0/SB8x0/SB9x0 USB EHCI Controller
                       ATI SB7x0/SB8x0/SB9x0 USB OHCI2 Controller
                       ATI SB7x0/SB8x0/SB9x0 USB EHCI Controller
                       ATI SB7x0/SB8x0/SB9x0 USB OHCI0 Controller
                       ATI SB7x0/SB8x0/SB9x0 USB OHCI0 Controller
bios:
                       BIOS
bridge:
                       ATI RD9x0/RX980 Host Bridge
                       AMD Family 15h Processor Function 4
                       ATI SB900 PCI to PCI bridge (PCIE port 2)
                       AMD Family 15h Processor Function 2
                       ATI SB700/SB800/SB900 PCI to PCI bridge (PCIE port 0)
                       AMD Family 15h Processor Function 0
                       ATI SBx00 PCI to PCI Bridge
                       ATI RD890/RD9x0/RX980 PCI to PCI bridge (PCI Express GPP Port 4)
                       ATI RD890/RD9x0/RX980 PCI to PCI bridge (PCI Express GFX port 0)
                       AMD Family 15h Processor Function 5
                       ATI SB900 PCI to PCI bridge (PCIE port 3)
                       AMD Family 15h Processor Function 3
                       AMD Family 15h Processor Function 1
                       ATI RD890/RD9x0/RX980 PCI to PCI bridge (PCI Express GPP Port 0)
                       ATI SB7x0/SB8x0/SB9x0 LPC host controller
hub:
                       Linux Foundation 1.1 root hub
                       Standard Microsystems Hub
                       Linux Foundation 3.0 root hub
                       Linux Foundation 1.1 root hub
                       Linux Foundation 3.0 root hub
                       Linux Foundation 1.1 root hub
                       Linux Foundation 2.0 root hub
                       Standard Microsystems Hub
                       Linux Foundation 2.0 root hub
                       Linux Foundation 1.1 root hub
                       Linux Foundation 2.0 root hub
                       Linux Foundation 2.0 root hub
                       Linux Foundation 2.0 root hub
memory:
                       Main Memory
unknown:
                       FPU
                       DMA controller
                       PIC
                       Keyboard controller
                       ATI SBx00 SMBus Controller
                       Serial controller
  /dev/input/event19   Microsoft Wired Keyboard 600 (model 1576)
  /dev/ttyUSB0         Texas Instruments TUSB3410 Microcontroller
  /dev/input/event2    Logitech G430 Surround Sound Gaming Headset
```

And the benchmark results [with this gist](https://gist.github.com/vltr/7965fac5e7db795c803137377ffd3674) (100k iteractions):

```
$ pip freeze                    
multidict==4.3.1
timy==0.4.0

$ python -B test_cidict_multidict.py
CIDict executed (test_cidict) for 100000 times in 52.994762 seconds
CIDict best time was 0.000488 seconds
multidict.CIMultiDict executed (test_cimultidict) for 100000 times in 19.562090 seconds
multidict.CIMultiDict best time was 0.000181 seconds

$ python -B test_cidict_multidict.py profile
         135000003 function calls in 85.840 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000   24.060    0.000   85.724    0.001 test_cidict_multidict.py:148(test_cidict)
 15900000   14.112    0.000   20.426    0.000 test_cidict_multidict.py:50(__contains__)
 15900000   13.844    0.000   20.276    0.000 test_cidict_multidict.py:44(__getitem__)
   600000    7.123    0.000   14.992    0.000 test_cidict_multidict.py:36(_convert)
 37100000    6.120    0.000    6.120    0.000 {method 'casefold' of 'str' objects}
  5300000    5.734    0.000    6.616    0.000 test_cidict_multidict.py:47(__setitem__)
 15900000    3.804    0.000    3.804    0.000 {function CIDict.__getitem__ at 0x7fb0c92e7b70}
 15900000    3.706    0.000    3.706    0.000 {function CIDict.__contains__ at 0x7fb0c92e7c80}
   600000    1.824    0.000   16.816    0.000 test_cidict_multidict.py:32(__init__)
 10600000    1.821    0.000    1.821    0.000 {method 'upper' of 'str' objects}
 10600000    1.729    0.000    1.729    0.000 {method 'lower' of 'str' objects}
  5300000    1.130    0.000    1.130    0.000 {method 'pop' of 'dict' objects}
   600000    0.594    0.000    0.594    0.000 test_cidict_multidict.py:151(<listcomp>)
   600000    0.123    0.000    0.123    0.000 {method 'keys' of 'dict' objects}
        1    0.117    0.117   85.840   85.840 <string>:1(<module>)
        1    0.000    0.000   85.840   85.840 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


         21900003 function calls in 23.874 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000   20.285    0.000   23.776    0.000 test_cidict_multidict.py:161(test_cimultidict)
 10600000    1.519    0.000    1.519    0.000 {method 'upper' of 'str' objects}
 10600000    1.458    0.000    1.458    0.000 {method 'lower' of 'str' objects}
   600000    0.515    0.000    0.515    0.000 test_cidict_multidict.py:164(<listcomp>)
        1    0.098    0.098   23.874   23.874 <string>:1(<module>)
        1    0.000    0.000   23.874   23.874 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

If there's anything else I can do, please, let me know :wink:

**Comment by vltr** [comment]
Oops, sorry, forgot the Python version and to test against Python 3.5. For the benchmark above:

```
$ python --version                          
Python 3.6.5
```

And for Python 3.5:

```
$ python --version          
Python 3.5.5

$ pip freeze                
multidict==4.3.1
timy==0.4.0

$ python -B test_cidict_multidict.py
CIDict executed (test_cidict) for 100000 times in 53.395700 seconds
CIDict best time was 0.000470 seconds
multidict.CIMultiDict executed (test_cimultidict) for 100000 times in 22.533207 seconds
multidict.CIMultiDict best time was 0.000196 seconds

$ python -B test_cidict_multidict.py profile
         135000003 function calls in 85.974 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000   26.802    0.000   85.836    0.001 test_cidict_multidict.py:148(test_cidict)
 15900000   14.049    0.000   18.981    0.000 test_cidict_multidict.py:44(__getitem__)
 15900000   13.794    0.000   18.939    0.000 test_cidict_multidict.py:50(__contains__)
   600000    7.868    0.000   15.700    0.000 test_cidict_multidict.py:36(_convert)
  5300000    5.785    0.000    6.845    0.000 test_cidict_multidict.py:47(__setitem__)
 37100000    5.147    0.000    5.147    0.000 {method 'casefold' of 'str' objects}
 15900000    3.085    0.000    3.085    0.000 {function CIDict.__contains__ at 0x7ff0c7acbc80}
 15900000    2.905    0.000    2.905    0.000 {function CIDict.__getitem__ at 0x7ff0c7acbb70}
   600000    1.870    0.000   17.571    0.000 test_cidict_multidict.py:32(__init__)
 10600000    1.453    0.000    1.453    0.000 {method 'upper' of 'str' objects}
 10600000    1.393    0.000    1.393    0.000 {method 'lower' of 'str' objects}
  5300000    0.895    0.000    0.895    0.000 {method 'pop' of 'dict' objects}
   600000    0.696    0.000    0.696    0.000 test_cidict_multidict.py:151(<listcomp>)
        1    0.139    0.139   85.974   85.974 <string>:1(<module>)
   600000    0.092    0.000    0.092    0.000 {method 'keys' of 'dict' objects}
        1    0.000    0.000   85.974   85.974 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


         21900003 function calls in 25.883 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000   22.737    0.000   25.771    0.000 test_cidict_multidict.py:161(test_cimultidict)
 10600000    1.230    0.000    1.230    0.000 {method 'lower' of 'str' objects}
 10600000    1.214    0.000    1.214    0.000 {method 'upper' of 'str' objects}
   600000    0.590    0.000    0.590    0.000 test_cidict_multidict.py:164(<listcomp>)
        1    0.112    0.112   25.883   25.883 <string>:1(<module>)
        1    0.000    0.000   25.883   25.883 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

I hope this helps.

File: sanic/cookies.py [hunk 2]
```diff
@@ -47,16 +47,15 @@ def __init__(self, headers):
...
     def __setitem__(self, key, value):
         # If this cookie doesn't exist, add it to the header keys
-        cookie_header = self.cookie_headers.get(key)
-        if not cookie_header:
+        if not self.cookie_headers.get(key):
             cookie = Cookie(key, value)
             cookie['path'] = '/'
-            cookie_header = MultiHeader("Set-Cookie")
-            self.cookie_headers[key] = cookie_header
-            self.headers[cookie_header] = cookie
```
Comment (yunstanford) [inline sanic/cookies.py]: `CIMultiDict` only supports `str` or `str` subclass as key.

File: sanic/cookies.py [hunk 3]
```diff
@@ -67,7 +66,12 @@ def __delitem__(self, key):
             self[key]['max-age'] = 0
         else:
             cookie_header = self.cookie_headers[key]
-            del self.headers[cookie_header]
+            # remove it from header
+            cookies = self.headers.popall(cookie_header)
```
Comment (yunstanford) [inline sanic/cookies.py]: workaround for deleting cookie

**Comment by yunstanford** [comment]
@r0fls Switched to `CIMultiDict`.

**Comment by r0fls** [comment]
@vltr regarding:

>I have no ties with this project (AFAIK, but I wish I could), so, why will I be blocking anything?

I just meant you can make a comment linking to the duplicate issue, if you have time, and suggesting to close the duplicate issue (or the original, if the duplicate has more information and/or discussion).

File: sanic/cookies.py [hunk 4]
```diff
@@ -47,16 +47,15 @@ def __init__(self, headers):
         super().__init__()
         self.headers = headers
         self.cookie_headers = {}
+        self.header_key = "Set-Cookie"
```
Comment (r0fls) [inline sanic/cookies.py]: Would it make more sense to call this `cookie_key`?

Comment (r0fls) [reply sanic/cookies.py]: Oh, it's in the cookies.py file, didn't notice that... probably fine as is.

File: sanic/cookies.py (hunk 3 above)
Comment (r0fls) [reply sanic/cookies.py]: Can we consolidate this to:

```
if cookie.key != key:
    self.headers.add(cookie_header, cookie)
```

File: sanic/cookies.py [hunk 5]
```diff
@@ -67,7 +66,12 @@ def __delitem__(self, key):
             self[key]['max-age'] = 0
         else:
             cookie_header = self.cookie_headers[key]
-            del self.headers[cookie_header]
+            # remove it from header
+            cookies = self.headers.popall(cookie_header)
+            for cookie in cookies:
+                if cookie.key == key:
```
Comment (r0fls) [inline sanic/cookies.py]: Can we consolidate this to:

```
if cookie.key != key:
    self.headers.add(cookie_header, cookie)
```

File: sanic/response.py [hunk 6]
```diff
@@ -53,7 +54,7 @@ def __init__(self, streaming_fn, status=200, headers=None,
         self.content_type = content_type
         self.streaming_fn = streaming_fn
         self.status = status
-        self.headers = headers or {}
+        self.headers = CIMultiDict(headers or {})
```
Comment (r0fls) [inline sanic/response.py]: I know you didn't do this, but would it make sense to just make `{}` the default for `headers`, instead of `None`?

**Comment by vltr** [comment]
@r0fls 

> I just meant you can make a comment linking to the duplicate issue, if you have time, and suggesting to close the duplicate issue (or the original, if the duplicate has more information and/or discussion).

Yes, sorry, #1242 if I recall. Also, I think PR #1131 might be a duplicate (regarding cookies).

**Comment by vltr** [comment]
To complement: there's a lot of interaction on these issues and PR as well. What I think is that someone needs to step up and start making some choices (if properly sustained) about how things will be with Sanic. Regarding cookies, `multidict` seems faster than the actual `CIDict` implementation, but does it cover all edges regarding cookie handling? Or we need to go with `werkzeug` or [`Headers`](https://docs.python.org/3/library/wsgiref.html#wsgiref.headers.Headers), as pointed by @pyx? I mean, I think these discussions needs to be centralized instead of "lost" in a sea of comments between PRs and issues ... My two cents: create a new milestone, attach new issues to it (like "How headers and cookies should be handled") and redirect all those discussions to that issue so, based on the community experience and feedback in that issue only, a proper, sustained decision can be done :simple_smile:

File: sanic/cookies.py (hunk 5 above)
Comment (yunstanford) [reply sanic/cookies.py]: sure

File: sanic/response.py (hunk 6 above)
Comment (yunstanford) [reply sanic/response.py]: No, python default argument is mutable, making `{}` default is a bad pattern. @r0fls 

http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

**Comment by yunstanford** [comment]
@vltr yeah, agree.

`multidict` should be a fully compatible replacement for the default headers (https://docs.python.org/3/library/wsgiref.html#wsgiref.headers.Headers) according to the doc.

Also, if we didn't cover all edges regarding cookie handling, it means we're missing test coverage for this part. We should add more unit tests around that.

**Comment by vltr** [comment]
@yunstanford 

> Also, if we didn't cover all edges regarding cookie handling, it means we're missing test coverage for this part. We should add more unit tests around that.

Yeap. I sincerely think there must be a "standard test" to validate this. I'll check it out later and drop a line here.

**Comment by vltr** [comment]
@yunstanford why not Python's own test cases? Check [here](https://github.com/python/cpython/blob/v3.7.0/Lib/test/test_wsgiref.py#L465), [here](https://github.com/python/cpython/blob/v3.7.0/Lib/test/test_httplib.py#L128), [here](https://github.com/python/cpython/blob/v3.7.0/Lib/test/test_http_cookies.py) and [here](https://github.com/python/cpython/blob/v3.7.0/Lib/test/test_http_cookiejar.py). I know it's overwhelming (the amount of tests), but they may provide some edge cases for testing ...

**Comment by yunstanford** [comment]
@vltr yep, thanks for bringing this up. those are great. But some of them should be already covered by `multidict`.

**Comment by vltr** [comment]
Yeap. It was more like to find some edge case here or there ... No way one should "copy" all these tests, like I said, it's overwhelming :sweat_smile:

**Comment by r0fls** [comment]
ok let's merge this for now, someone is welcome to provide a better implementation if there is one
